src/aspire/classification/align2d.py

```diff
@@ -1,66 +1,188 @@
+        :param basis_coefficients: (n_img, self.pca_basis.count) compressed basis coefficients
+
+        :returns: Image instance (stack of images)
+        """
+
+    def _cls_images(self, cls, src=None):
+        """
+        Util to return images as an array for class k (provided as array `cls` ),
+        preserving the class/nbor order.
+
+        :param cls: An iterable (0/1-D array or list) that holds the indices of images to align.
+        In Class Averaging, this would be a class.
+        :param src: Optionally overridee the src, for example, if you want to use a different
+        source for a certain operation (ie aignment).
+        """
+        src = src or self.src
+
+        n_nbor = cls.shape[-1]  # Includes zero'th neighbor
+
+        images = np.empty((n_nbor, src.L, src.L), dtype=self.dtype)
+
+        for i, index in enumerate(cls):
+            images[i] = src.images(index, 1).asnumpy()
+
+        return images
+
 
-        :returns: (classes, reflections, rotations, shifts, correlations)
+class AveragedAlign2D(Align2D):
+    """
+    Subclass supporting aligners which perform averaging during output.
+    """
+
+    def align(self, classes, reflections, basis_coefficients):
+        """
+        See Align2D.align
+        """
+        # Correlations are currently unused, but left for future extensions.
+        cls, ref, rot, shf, corrs = self._align(
+            classes, reflections, basis_coefficients
+        )
+        return self.average(cls, ref, rot, shf), cls, ref, rot, shf, corrs
+
+    def average(
+        self,
+        classes,
+        reflections,
+        rotations,
+        shifts=None,
+        coefs=None,
+    ):
+        """
+        Combines images using averaging in `self.composite_basis`.
+
+        :param classes: class indices (refering to src). (n_img, n_nbor)
+        :param reflections: Bool representing whether to reflect image in `classes`
+        :param rotations: Array of in-plane rotation angles (Radians) of image in `classes`
+        :param shifts: Optional array of shifts for image in `classes`.
+        :coefs: Optional Fourier bessel coefs (avoids recomputing).
+        :return: Stack of Synthetic Class Average images as Image instance.
         """
-        raise NotImplementedError("Subclasses must implement align.")
+        n_classes, n_nbor = classes.shape
+
+        b_avgs = np.empty((n_classes, self.composite_basis.count), dtype=self.src.dtype)
+
+        for i in tqdm(range(n_classes)):
+
+            # Get coefs in Composite_Basis if not provided as an argumen.
+            if coefs is None:
+                # Retrieve relavent images directly from source.
+                neighbors_imgs = Image(self._cls_images(classes[i]))
+
+                # Do shifts
+                if shifts is not None:
+                    neighbors_imgs.shift(shifts[i])
+
+                neighbors_coefs = self.composite_basis.evaluate_t(neighbors_imgs)
```

Comment:
I'll try it and see what happens. No warranty.

Reply:
It was terribly slow, so I reverted it. Think might need to circle back on that one.

Reply:
Ok. Normally, `expand` should be slower than `evaluate_t`, but not more than 3–4×. If it's slower, there are convergence issues somewhere that we need to look at.

Reply:
Looks like a factor of 3x, which is pretty slow, but I'll leave it based on your comment this is expected/acceptable.

Reply:
Thanks. This is an area where some optimization may be helpful but unfortunately, we can never expect to do be as fast as `evaluate_t`. I doubt that we can actually get the ratio faster than 3× (this depends on the condition number of the basis), but we may be able to speed both up by optimizing `evaluate_t`.

Reply:
I tried to leave it, I really did, but it is just not reasonable. Unit tests, even small ones, will potentially time out and make simple development totally obnoxious.  I will make an issue for someone to improve the performance and string replace `evaluate_t` with `expand`.  This PR is going to keep `evaluate_t` as we have been using for everything else since I worked here. Now is not the time...

Reply:
It's fine to replace `expand` with `evaluate_t` for tests, but in the library internals, it's problematic. I see we use it this way in at least three other places (`RIRClass2D`,`DenoiserCov2D`, and `CtfEstimator`).

We will leave it for now and I'll make an issue.